src/neo-vm/ExecutionEngine.cs

```diff
@@ -1154,6 +1194,84 @@ private bool ExecuteInstruction()
             return true;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool ExecuteTry(int catchOffset, int finallyOffset)
+        {
+            if (catchOffset < 0 || finallyOffset < 0) return false;
+            if (finallyOffset + catchOffset <= 0) return false;
+            if (finallyOffset > 0 && catchOffset >= finallyOffset) return false;
+
+            if (CurrentContext.ErrorHandle == null)
+                CurrentContext.ErrorHandle = new ErrorHandle();
+            CurrentContext.ErrorHandle.Push(new TryContext(CurrentContext, catchOffset, finallyOffset));
+            CurrentContext.InstructionPointer += CurrentContext.CurrentInstruction.Size;
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal bool ExecuteEndTryCatch(TryState currentState)
+        {
+            var currentTry = CurrentContext.ErrorHandle?.CurContext;
+            if (currentTry is null) return false;
+            if (currentTry.State != currentState) return false;
+
+            if (currentTry.HasFinally)
+            {
+                currentTry.State = TryState.Finally;
+                currentTry.EndTryCatch(CurrentContext.InstructionPointer + CurrentContext.CurrentInstruction.Size);
+                CurrentContext.InstructionPointer = currentTry.FinallyPointer;
+            }
+            else
+            {
+                CurrentContext.ErrorHandle.Pop();
+                int nextOpcodePos = checked(CurrentContext.InstructionPointer + CurrentContext.CurrentInstruction.Size);
+                CurrentContext = InvocationStack.Peek();
+                CurrentContext.InstructionPointer = nextOpcodePos;
+            }
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool ExecuteEndFinally()
+        {
+            var currentTry = CurrentContext.ErrorHandle?.Pop();
+            if (currentTry is null) return false;
+
+            if (FaultState.Rethrow)
+            {
+                FaultState.Rethrow = false;
+                State = VMState.FAULT;
+                return false;
+            }
+            CurrentContext.InstructionPointer = currentTry.EndPointer;
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool ExecuteThrow(string errorinfo)
+        {
+            FaultState.Error = new CatcheableError(errorinfo);
+            State = VMState.FAULT;
+            return false;
+        }
+
+        private bool HandleError()
+        {
+            for (var i = 0; i < InvocationStack.Count; i++)
+            {
+                var content = InvocationStack.ElementAt(i);
+                if (content.ErrorHandle != null)
+                {
+                    if (content.ErrorHandle.HandleError(this))
+                    {
+                        State = VMState.NONE;
```

Comment:
We need to clean the error?

```suggestion
                        State = VMState.NONE;
                        FaultState.IsCatcheableError = false;
```

Reply:
Yes, we need to clear the error, but I think it shoud be cleared in `endcatch` body.

Reply:
Applied, plz have a review again.